enterprise_access/apps/content_assignments/models.py

```diff
@@ -85,6 +92,137 @@ def policy(self):
         except ObjectDoesNotExist:
             return None
 
+    def _should_acknowledge_expired_assignment(self, assignment):
+        """
+        Returns a tuple of booleans indicating whether the given assignment should be acknowledged and
+        whether the assignment has already been acknowledged.
+        """
+        if assignment.state != LearnerContentAssignmentStateChoices.EXPIRED:
+            return False, False
+
+        last_expiration = assignment.get_last_successful_expiration_action()
+        expiration_last_acknowledged = assignment.get_last_successful_acknowledged_expired_action()
+        if not last_expiration:
+            logger.error(
+                'Assignment %s is in state %s but has no successful expiration action.',
+                assignment.uuid,
+                LearnerContentAssignmentStateChoices.EXPIRED,
+            )
+
+        # Check whether expiration has ever been acknowledged; if not, acknowledge it.
+        if not expiration_last_acknowledged:
+            return True, False
+
+        # If it has been acknowledged before, check whether last expiration action is newer
+        # than the last acknowledged expiration action.
+        if last_expiration and last_expiration.completed_at > expiration_last_acknowledged.completed_at:
+            return True, True
```

Comment:
I have a sneaking suspicion that this case can be lumped into the previous one (return True, False), and this whole method would only have 3 possible outputs: (False, False), (True, False), and (False, True).  Logically, it makes sense to me that if there's been a more recent expiration than the latest expiration acknowledgement, then that really means the latest acknowledgement has been invalidated and there's effectively no existing acknowledgement.

Later in acknowledge_assignments() there is no logic which distinguishes between (True, False) and (True, True) anyway, and the absence of the (True, True) case can simplify the definitions of `has_already_acknowledged_expiration`/`has_already_acknowledged_cancellation`.